### PR TITLE
Keep device extensions in renewed sessions

### PR DESCRIPTION
**Pull request by codingllama**
Renewal occurs via `/webapi/sessions/web/renew`, which after a few steps ends up at Server.ExtendWebSession.

I've added the WebSessionSpecV2.HasDeviceExtensions field to avoid needlessly parsing certificates for every renewal (it may come in handy for the Web UI too). I've also made Server.NewWebSession private, as it now takes parameters that shouldn't be visible or used outside of lib/auth.

https://github.com/gravitational/teleport.e/issues/3236

### CODE REVIEW

**Comment by codingllama** [comment]
PR should be fairly small apart from generated protos.

FYI @avatus, we could use HasDeviceExtensions to easily check that in the Web API.

**Comment by codingllama** [comment]
Conflicts are due to #39977. Easy to fix.

**Review by probakowski — APPROVED** [review]

File: lib/auth/auth.go [hunk 1]
```diff
@@ -3997,6 +4008,28 @@ func (a *Server) ExtendWebSession(ctx context.Context, req WebSessionReq, identi
 	return sess, nil
 }
 
+func decodeDeviceExtensionsFromSession(webSession types.WebSession) (*tlsca.DeviceExtensions, error) {
+	// Reading the extensions from the session itself means we are always taking
+	// them for a legitimate source (ie, certificates issued by Auth).
```
Comment (ibeckermayer) [inline lib/auth/auth.go:4013]: Therefore we needn't validate the cert here? Consider clarifying

File: lib/auth/auth.go [hunk 2]
```diff
@@ -4582,8 +4615,22 @@ func (a *Server) GetTokens(ctx context.Context, opts ...services.MarshalOption)
...
 
-// NewWebSession creates and returns a new web session for the specified request
-func (a *Server) NewWebSession(ctx context.Context, req types.NewWebSessionRequest) (types.WebSession, error) {
+// newWebSessionOpts are WebSession creation options exclusive to Auth.
+// These options complement [types.NewWebSessionRequest].
+// See [Server.newWebSession].
+type newWebSessionOpts struct {
+	// deviceExtensions are the device extensions to apply to the session.
+	// Only present on renewals, the original extensions are applied by
+	// [Server.AugmentWebSessionCertificates].
```
Comment (ibeckermayer) [inline lib/auth/auth.go:4624]: Any sensible way to consolidate this? As in, have `AugmentWebSessionCertificates` and `newWebSession` call the same thing? I don't understand this section of the code well, just a "spidey sense" that maybe you could simplify something here, could be totally wrong though.

**Review by ibeckermayer — APPROVED** [review]
Approved but see comments.

File: lib/auth/auth.go (hunk 1 above)
Comment (codingllama) [reply lib/auth/auth.go:4013]: Correct, I'll clarify.

File: lib/auth/auth.go (hunk 2 above)
Comment (codingllama) [reply lib/auth/auth.go:4624]: This is used as the input for `a.generateUserCert`, so there isn't much more to simplify. AugmentWebSessionCertificates does a whole lot of verifications before letting the extensions end up in the certificate, but during renewal we just pass them along. Other kinds of device trust renewals (think MFA certificates) behave similarly.

I hope this clarifies a bit.

**Comment by codingllama** [comment]
Rebased, conflicts solved (deleted "PreAuthenticated" method and the external "move" of types.NewWebSessionRequest to auth.NewWebSessionRequest), comments addressed.

**Pull request merged**